tla-assignments/src/main/scala/at/forsyte/apalache/tla/assignments/SmtFreeSymbolicTransitionExtractor.scala

```diff
@@ -0,0 +1,174 @@
+  private case class PartialState( unassignedVars: Set[String], partialStrat: StrategyType )
+
+  // Since disjunction and ITE behave exactly the same w.r.t. assignments, we write a unified routine
+  private def handleDisjunctionOrITE( s: PartialState, letInOperMap: BodyMap , args: Seq[TlaEx] ): PartialState = {
+    // We independently process each disjunct
+    val childStates = args map { getStrategyOptInternal(s, letInOperMap)(_) }
+    val unassignedVarsSeq = childStates map { _.unassignedVars }
+    // We need to see whether all branches assign exactly the same variables
+    val unassignedEverywhere = unassignedVarsSeq.foldLeft( s.unassignedVars ){ _.intersect( _ ) }
+    val unassignedSomewhere = unassignedVarsSeq.foldLeft( Set.empty[String] ) { _.union( _ ) }
+    // Variables that are unassigned on at least one branch and assigned on all others
+    val imbalancedAssignments = unassignedSomewhere -- unassignedEverywhere
+    if (imbalancedAssignments.nonEmpty) {
+      // Report problms for each disjunct, if any
+      val problems : Seq[String] = args.zip( childStates ) flatMap {
+        case (childEx, PartialState( childUnassignedVars, _ )) =>
+          val s = childUnassignedVars.intersect( imbalancedAssignments )
+          if ( s.isEmpty ) None
+          else {
+            val locString = getLocString( childEx )
+            Some( s"$locString - Missing assignments to: ${s.mkString( "," )}" )
+          }
+      }
+
+      throw new AssignmentException( problems.mkString("\n") )
+    }
+    // Assuming all branches assign the same variables, works correctly if args.isEmpty (for whatever reason)
+    val unifiedStrat = (childStates map { _.partialStrat }).foldLeft( s.partialStrat ) { _ ++ _ }
+    PartialState( unassignedEverywhere, unifiedStrat )
+  }
+
+  // Transition method between partial states
+  private def getStrategyOptInternal( s: PartialState, letInOperMap: BodyMap ): TlaEx => PartialState = {
+    /** Base case, assignment candidates  */
+    case ex@OperEx( TlaOper.eq, OperEx( TlaActionOper.prime, NameEx( name ) ), star ) =>
+      // First, check if star contains unassigned varaible access
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      // if `name` is not yet assigned, this spot becomes an assignment
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else s
+
+    /** Base case, manual assignments */
+    case ex@OperEx( BmcOper.assign, OperEx( TlaActionOper.prime, NameEx( name ) ), star ) =>
+      // Similar to above, except manual assignments throw if spurious
+      throwOnUseBeforeAssignment( s.unassignedVars )( star )
+      if ( s.unassignedVars.contains( name )) PartialState( s.unassignedVars - name, s.partialStrat :+ ex.ID )
+      else {
+        val locString = getLocString( ex )
+        throw new AssignmentException( s"Manual assignment at $locString is spurious: $name is already assigned!" )
```

Comment:
```suggestion
        throw new AssignmentException( s"Manual assignment at $locString is spurious: $name is already assigned!" )
```
```suggestion
        throw new AssignmentException( s"$locString: manual assignment is spurious: $name is already assigned!" )
```

Reply:
If there is a specific format we want for these kinds of messages, then I suggest we create a function to generate the desired formatting rather than having to enforce it manually. (This would be for a separate PR obviously).